lib/kube/proxy/transport_retryable.go

```diff
@@ -0,0 +1,274 @@
+			rt.log.DebugContext(req.Context(),
+				"Unable to close request body",
+				"size", req.ContentLength,
+				"method", req.Method,
+				"path", req.URL.Path,
+				"error", err,
+			)
+		}
+		rt.semaphore.Release(req.ContentLength)
+	}
+	return releaseSemaphore
+}
+
+// isUpgrade detects HTTP/1.1 protocol upgrades (WebSocket, SPDY) that won't
+// receive HTTP/2 GOAWAY frames.
+//
+// We cannot check req.Proto to detect HTTP/2 because modifyRequest() sets it
+// to "HTTP/1.1" for all outgoing requests. Instead, we rely on the fact that
+// all transports wrapped by retryableTransport are created via newH2Transport(),
+// which configures HTTP/2. Requests without Upgrade headers will use HTTP/2 and
+// may receive GOAWAY.
+func (rt *retryableTransport) isUpgrade(req *http.Request) bool {
+	return req.Header.Get("Connection") == "Upgrade" ||
+		req.Header.Get("Upgrade") != "" ||
+		req.Header.Get("X-Stream-Protocol-Version") != ""
+}
+
+// CloseIdleConnections closes idle connections in the inner transport.
+func (rt *retryableTransport) CloseIdleConnections() {
+	if ci, ok := rt.inner.(interface{ CloseIdleConnections() }); ok {
+		ci.CloseIdleConnections()
+	}
+}
+
+// retryBuffer incrementally reads from a source, incrementally writes
+// to a buffer, and gracefully completes reading during any early call to Close.
+type retryBuffer struct {
+	src     io.ReadCloser
+	buf     *bytes.Buffer
+	mu      sync.Mutex
+	cond    *sync.Cond
+	readErr error
+}
+
+func newRetryBuffer(source io.ReadCloser, capacity int) *retryBuffer {
+	type limitedReadCloser struct {
+		*io.LimitedReader
+		io.Closer
+	}
+	rb := &retryBuffer{
+		// Limit reading in case actual body size is
+		// greater than presented ContentLength.
+		src: &limitedReadCloser{
+			LimitedReader: &io.LimitedReader{R: source, N: int64(capacity)},
+			Closer:        source,
+		},
+		buf: bytes.NewBuffer(make([]byte, 0, capacity)),
+	}
+	rb.cond = sync.NewCond(&rb.mu)
+	return rb
+}
+
+func (rb *retryBuffer) Read(p []byte) (int, error) {
+	rb.mu.Lock()
+	defer rb.mu.Unlock()
+
+	if rb.src == nil {
+		return 0, io.EOF
+	}
+
+	n, err := rb.src.Read(p)
+	if n > 0 {
+		rb.buf.Write(p[:n])
+	}
+	if err != nil && !errors.Is(err, io.EOF) && rb.readErr == nil {
+		rb.readErr = err
+		rb.cond.Broadcast()
+	}
+	return n, err
+}
+
+// Close completes buffering by reading any remaining data,
+// ensuring getBody() always has the complete buffer.
+func (rb *retryBuffer) Close() error {
+	rb.mu.Lock()
+	defer rb.mu.Unlock()
+
+	// Close is idempotent.
+	if rb.src == nil {
+		return nil
+	}
+
+	// If the connection closed in the middle of sending
+	// due to a GOAWAY, read remaining data for a retry attempt.
+	remaining, readErr := io.ReadAll(rb.src)
+	if readErr != nil && !errors.Is(readErr, io.EOF) && rb.readErr == nil {
+		rb.readErr = readErr
+	}
+	if len(remaining) > 0 {
+		rb.buf.Write(remaining)
+	}
+
+	// Close source and mark as done.
+	err := rb.src.Close()
+	rb.src = nil
+	rb.cond.Broadcast()
```

Comment:
We don't need to read the full request here, and we definitely must not not block Close on I/O from the original client while we buffer the request, especially because we might not retry at all, depending on why the body is getting closed before reaching the end. The second body can just read from the buffer until the end of the buffer, then read (and buffer) from the source.